portal-kernel/src/com/liferay/portal/kernel/internal/service/persistence/TableMapperImpl.java

```diff
@@ -444,12 +447,23 @@ protected static <T extends BaseModel<T>> List<T> getBaseModels(
 			return Collections.emptyList();
 		}
 
+		long currentCompanyId = CompanyThreadLocal.getCompanyId();
 		List<T> slaveBaseModels = new ArrayList<>(slavePrimaryKeys.length);
 
 		try {
 			for (long slavePrimaryKey : slavePrimaryKeys) {
-				slaveBaseModels.add(
-					slaveBasePersistence.findByPrimaryKey(slavePrimaryKey));
+				T slaveBaseModel =
+					slaveBasePersistence.findByPrimaryKey(slavePrimaryKey);
+
+				long checkableCompanyId =
+					_getCheckableCompanyId(slaveBaseModel);
+
+				boolean isSkipCheck =
+					checkableCompanyId == CompanyConstants.SYSTEM;
+
+				if (isSkipCheck || checkableCompanyId == currentCompanyId) {
```

Comment:
I think, with your logic you won't never return all the records if CompanyThreadLocal is System.  Am I right? That would break some internal processes where the CompanyThreadLocal is not necessary. 

Would this work for you?
`currentCompanyId == CompanyConstants.SYSTEM || currentCompanyId == _getCheckableCompanyId(slaveBaseModel)`

Reply:
@gianmarcobrunialti

Reply:
Yes, but if CompanyThreadLocal.getCompanyId() *and* _getCheckabeCompanyId(slaveBaseModel) *both return* `SYSTEM` for the current iteration, which should be the correct case. Of course the code you suggest makes it quicker to read! :)

Reply:
Let's re-evaluate everything to be sure:
```
boolean isSkipCheck = checkableCompanyId == CompanyConstants.SYSTEM;

if (isSkipCheck || checkableCompanyId == currentCompanyId) {
```
- Case currentCompanyId == SYSTEM: we will only returns the records with companyId SYSTEM
- Case currentCompanyId != SYSTEM: we will return the records with companyId XXX or SYSTEM

```
if ((currentCompanyId == CompanyConstants.SYSTEM) || (currentCompanyId == _getCheckableCompanyId(slaveBaseModel))
```
- Case currentCompanyId == SYSTEM: we will return all related records
- Case currentCompanyId != SYSTEM: we will return only the records with companyId XXX

The first solution is more restrictive (or safer) but it won't return any record (since the SYSTEM company is not persisted) when CompanyThreadLocal is SYSTEM and that will break upgrades or internal processes.

Or we use the second option assuming that (which it is not too bad since with maybe makes sense for SYSTEM cases) or we have a way to get the companyId from the masterModel.

Let me know what you think about this.

Reply:
All clear, the second option makes it then, we basically reintroduce SYSTEM cases. Great, I'm about to push 👍🏼